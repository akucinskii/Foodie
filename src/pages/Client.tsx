import React, { useEffect } from "react";
import Button from "../components/Button";
import { McList } from "../utils/McList";

type McListType = typeof McList;
type McListItemType = typeof McList[number];

const Client = () => {
  const [McListState, setMcListState] = React.useState<McListType>(McList);
  const [order, setOrder] = React.useState<McListItemType[]>([]);
  const [total, setTotal] = React.useState<number>(0);

  useEffect(() => {
    const initialValue = 0;
    const value = order.reduce(
      (prev, curr) => prev + curr.price * curr.quantity,
      initialValue
    );
    setTotal(value);
  }, [order]);

  const addToOrder = (item: McListItemType) => {
    const found = order.some((el) => el.name === item.name);

    const updatedMcList = McListState.map((el) => {
      if (el.name === item.name) {
        return { ...el, quantity: el.quantity + 1 };
      }
      return el;
    });

    setMcListState(updatedMcList);

    if (found) {
      const newOrder = order.map((el) => {
        if (el.name === item.name) {
          return { ...el, quantity: el.quantity + 1 };
        }
        return el;
      });

      setOrder(newOrder);
    } else {
      setOrder([...order, { ...item, quantity: 1 }]);
    }
  };

  const removeFromOrder = (item: McListItemType) => {
    const found = order.some((el) => el.name === item.name);

    const updatedMcList = McListState.map((el) => {
      if (el.quantity === 0) {
        return el;
      }

      if (el.name === item.name) {
        return { ...el, quantity: el.quantity - 1 };
      }
      return el;
    });

    setMcListState(updatedMcList);

    if (found) {
      const newOrder = order.map((el) => {
        if (el.name === item.name) {
          return { ...el, quantity: el.quantity - 1 };
        }
        return el;
      });

      setOrder(newOrder);

      if (item.quantity === 1) {
        const filteredOrder = order.filter((el) => el.name !== item.name);

        setOrder(filteredOrder);
      }
    }
  };

  const Form = McListState.map((item) => {
    return (
      <div className="flex gap-2" key={item.id}>
        <Button
          onClick={() => {
            addToOrder(item);
          }}
        >
          +
        </Button>
        <Button
          color="blue"
          onClick={() => {
            removeFromOrder(item);
          }}
        >
          -
        </Button>
        {item.name} {item.quantity}
      </div>
    );
  });
  return (
    <div className="grid gap-8 sm:grid-cols-2">
      <div className="flex max-w-sm flex-col justify-center gap-2">
        {Form}
        <Button
          color="blue"
          onClick={() => {
            console.log(order);
          }}
        >
          Submit
        </Button>
      </div>
      <div className="flex flex-col bg-slate-400 p-4">
        {order.map((item) => {
          return (
            <div className="flex gap-2" key={item.id}>
              {item.name} - {item.quantity}
            </div>
          );
        })}
        <div className="h-full" />
        <div className="bg-slate-500">Total {total}</div>
      </div>
    </div>
  );
};

export default Client;
