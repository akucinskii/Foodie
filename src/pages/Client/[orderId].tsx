import { useRouter } from "next/router";
import React, { useEffect } from "react";
import Button from "../../components/Button";
import { useSubmitOrderSlice } from "../../hooks/useSubmitOrderSlice";
import { McList } from "../../utils/McList";

export type McListType = typeof McList;
export type McListItemType = typeof McList[number];

const Client = () => {
  const [author, setAuthor] = React.useState<string>("");
  const [McListState, setMcListState] = React.useState<McListType>(McList);
  const [order, setOrder] = React.useState<McListItemType[]>([]);
  const [total, setTotal] = React.useState<number>(0);
  const router = useRouter();
  const orderId = router.query.orderId as string;

  const submitOrderSlice = useSubmitOrderSlice();

  useEffect(() => {
    const initialValue = 0;
    const value = order.reduce(
      (prev, curr) => prev + curr.price * curr.quantity,
      initialValue
    );
    setTotal(value);
  }, [order]);

  const addToOrder = (item: McListItemType) => {
    const found = order.some((el) => el.name === item.name);

    const updatedMcList = McListState.map((el) => {
      if (el.name === item.name) {
        return { ...el, quantity: el.quantity + 1 };
      }
      return el;
    });

    setMcListState(updatedMcList);

    if (found) {
      const newOrder = order.map((el) => {
        if (el.name === item.name) {
          return { ...el, quantity: el.quantity + 1 };
        }
        return el;
      });

      setOrder(newOrder);
    } else {
      setOrder([...order, { ...item, quantity: 1 }]);
    }
  };

  const removeFromOrder = (item: McListItemType) => {
    const found = order.some((el) => el.name === item.name);

    const updatedMcList = McListState.map((el) => {
      if (el.quantity === 0) {
        return el;
      }

      if (el.name === item.name) {
        return { ...el, quantity: el.quantity - 1 };
      }
      return el;
    });

    setMcListState(updatedMcList);

    if (found) {
      const newOrder = order.map((el) => {
        if (el.name === item.name) {
          return { ...el, quantity: el.quantity - 1 };
        }
        return el;
      });

      setOrder(newOrder);

      if (item.quantity === 1) {
        const filteredOrder = order.filter((el) => el.name !== item.name);

        setOrder(filteredOrder);
      }
    }
  };

  const Form = McListState.map((item) => {
    return (
      <div className="flex gap-2" key={item.id}>
        <Button
          onClick={() => {
            addToOrder(item);
          }}
        >
          +
        </Button>
        <Button
          onClick={() => {
            removeFromOrder(item);
          }}
        >
          -
        </Button>
        {item.name} {item.price}pln
      </div>
    );
  });
  return (
    <div className="flex flex-col gap-4">
      <h1 className="text-xl ">Warning: Please be sure about your order there is no edit/delete (yet)</h1>
      <div>
        <label className="label">
          <span className="label-text">Enter your name</span>
        </label>
        <input
          className="input input-bordered w-full"
          type="text"
          required
          placeholder="Author"
          value={author}
          onChange={(e) => setAuthor(e.target.value)}
        />
      </div>

      <div className="h-px flex-grow bg-gray-200"></div>

      <div className="grid gap-8 sm:grid-cols-2">
        <div className="flex max-w-sm flex-col justify-center gap-2">
          {Form}
        </div>

        <div className="flex flex-col gap-4">
          <div className="h-full overflow-x-auto">
            <table className="table-zebra table w-full">
              <thead>
                <tr>
                  <th>Item</th>
                  <th>Quantity</th>
                </tr>
              </thead>
              <tbody>
                {order.map((item) => {
                  return (
                    <tr key={item.id}>
                      <td className="flex gap-2">{item.name}</td>
                      <td> {item.quantity}</td>
                    </tr>
                  );
                })}
              </tbody>
            </table>
          </div>

          <div className=" sm:text:sm inline-block content-center items-center justify-center rounded px-6 py-2 text-center align-middle text-xs font-medium uppercase leading-tight text-black md:text-xl">
            Total: <span className="text-yellow-500">{total} pln.</span>
          </div>
          <Button
            disabled={author.length < 3 || order.length === 0}
            onClick={() => {
              submitOrderSlice(order, orderId, author);
              router.push(`/Driver/${orderId}`);
            }}
          >
            Submit
          </Button>
        </div>
      </div>
    </div>
  );
};

export default Client;
