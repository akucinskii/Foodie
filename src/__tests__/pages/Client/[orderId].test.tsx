import { render } from "@testing-library/react";
import { Session } from "next-auth";
import { describe, expect, it, vi, Mock } from "vitest";
import { signIn, signOut, useSession } from "next-auth/react";
import Client from "../../../pages/Client/[orderId]";
vi.mock("../../../utils/trpc", () => ({
  trpc: {
    useContext: () => ({}),
    order: {
      invalidate: vi.fn(),
      createOrderSlice: {
        useMutation: () => [vi.fn(), { isLoading: false }],
      },
    },
  },
  useQuery: () => ({
    data: {
      id: "1",
      name: "Test",
      email: "test",
    },
  }),
}));
vi.mock("next-auth/react");
vi.mock("next/router", () => ({
  useRouter: vi.fn(() => ({
    query: {
      orderId: "1",
    },
  })),
}));
describe("Navbar", () => {
  const mockUseSession = useSession as Mock;
  (signIn as Mock).mockImplementation(() => vi.fn());
  (signOut as Mock).mockImplementation(() => vi.fn());

  const renderComponent = () => {
    return render(<Client />);
  };

  const mockSession: Session = {
    user: {
      id: "1",
      name: "Test",
      email: "test@test.test",
      image: "/test",
    },
    expires: "1",
  };

  it("should match snapshot", () => {
    mockUseSession.mockReturnValue({
      data: mockSession,
      status: "unauthenticated",
    });

    const { container } = renderComponent();

    expect(container).toMatchSnapshot("logout");
  });
});
