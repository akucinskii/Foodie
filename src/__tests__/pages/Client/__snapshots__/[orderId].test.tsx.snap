// Vitest Snapshot v1

exports[`Navbar > should match snapshot > logout 1`] = `
<div>
  <div
    class="flex flex-col gap-4"
  >
    <h1
      class="text-xl "
    >
      Warning: Please be sure about your order. There is no edit/delete (yet)
    </h1>
    <div
      class="h-px flex-grow bg-gray-200"
    />
    <div
      class="grid gap-8 sm:grid-cols-2"
    >
      <div
        class="flex max-w-sm flex-col justify-center gap-2"
      >
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          BigMac
           
          8
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          McChicken
           
          8
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          McNuggets
           
          8
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          Wieśmac
           
          8
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          McRoyal
           
          8
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          McWrap
           
          8
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          McFlurry
           
          5
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          McShake wanilia
           
          5
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          McShake truskawka
           
          5
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          McShake czekolada
           
          5
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          kawa
           
          5
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          Zestaw BigMac
           
          14
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          Zestaw Mcwrap
           
          14
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          Zestaw McChicken
           
          14
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          Zestaw McNuggets
           
          14
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          Zestaw McRoyal
           
          14
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          Zestaw WieśMac
           
          14
          pln
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            -
          </button>
          <button
            class="btn btn-primary text-white"
            data-testid="button"
            type="button"
          >
            +
          </button>
          fryty
           
          5
          pln
        </div>
      </div>
      <div
        class="flex flex-col gap-4"
      >
        <div
          class="h-full overflow-x-auto"
        >
          <table
            class="table-zebra table w-full"
          >
            <thead>
              <tr>
                <th>
                  Item
                </th>
                <th>
                  Quantity
                </th>
              </tr>
            </thead>
            <tbody />
          </table>
        </div>
        <div
          class=" sm:text:sm inline-block content-center items-center justify-center rounded px-6 py-2 text-center align-middle text-xs font-medium uppercase leading-tight text-black md:text-xl"
        >
          Total: 
          <span
            class="text-yellow-500"
          >
            0
             pln.
          </span>
        </div>
        <button
          class="btn btn-primary text-white"
          data-testid="button"
          disabled=""
          type="button"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</div>
`;
